File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/template/abstract_class.dart
abstract class FormTemplate {
  /// Orchestrates the submit flow. Returns a success message or throws on error.
  Future<String> submit(Map<String, dynamic> formData) async {
    final cleanData = validate(formData);

    final result = await send(cleanData);

    await afterSuccess(result);

    return successMessage(result);
  }

  Map<String, dynamic> validate(Map<String, dynamic> data);

  Future<dynamic> send(Map<String, dynamic> cleanData);

  Future<void> afterSuccess(dynamic result);

  String successMessage(dynamic result);
}

File: implementations/react/src/patterns/behavioral/template/abstract-class.ts
/**
 * Clase Abstracta que define el esqueleto del algoritmo para manejar formularios.
 */

export abstract class FormTemplate {
  async submit(formData: Record<string, any>) {
    const cleanData = this.validate(formData);

    const result = await this.send(cleanData);

    await this.afterSuccess(result);

    return this.successMessage(result);
  }

  protected abstract validate(data: Record<string, any>): Record<string, any>;
  protected abstract send(cleanData: Record<string, any>): Promise<any>;
  protected abstract afterSuccess(result: any): Promise<void>;
  protected abstract successMessage(result: any): string;
}
